code/modules/power/stirling.dm

```diff
@@ -0,0 +1,224 @@
+
+	// The cycle requires a minimum temperature to overcome friction.
+	if(abs(air1.temperature - air2.temperature) < MIN_DELTA_T)
+		if(skipped_cycle) // Allow some leeway since networks can take some time to update.
+			stop_engine()
+		skipped_cycle = TRUE
+		return
+	skipped_cycle = FALSE
+
+	// Positive/negative Work from volume expansion/compression. Maximum volume is 1.5 tank volume.	The volume of the tank is not actually adjusted.
+	var/work_coefficient = working_volume.get_total_moles()*R_IDEAL_GAS_EQUATION*log(1.5)
+	var/work_done = work_coefficient*abs(air1.temperature - air2.temperature)*cycle_frequency
+
+	// Direction of heat flow, 1 for air1 -> air 2, -1 for air2 -> air 1
+	var/heat_dir = air1.temperature > air2.temperature ? 1 : -1
+
+	var/air1_dq = -heat_dir*work_coefficient*air1.temperature*cycle_frequency
+	var/air2_dq = heat_dir*work_coefficient*air2.temperature*cycle_frequency
+
+	var/power_generated = 0.75*work_done
+	// Excessive power is transferred as heat to the cooler side, reducing efficiency.
+	if(power_generated > max_power)
+		if(heat_dir == 1)
+			air2_dq += 0.75*(max_power - power_generated)
+		else
+			air1_dq += 0.75*(max_power - power_generated)
+		power_generated = max_power
+		if(prob(5))
+			spark_at(src, cardinal_only = TRUE)
+
+	air1.add_thermal_energy(air1_dq)
+	air2.add_thermal_energy(air2_dq)
+	
+	generate_power(power_generated)
+	last_gen = power_generated
+	genlev = max(0, min(round(4*power_generated / max_power), 4))
+	if(genlev != last_genlev)
+		update_icon()
+		update_sound()
+	last_genlev = genlev
+	update_networks()
+
+/obj/machinery/atmospherics/binary/stirling/examine(mob/user, distance)
+	. = ..()
+	if(distance > 1)
+		return
+
+	to_chat(user, "\The [src] is generating [round(last_gen/1000, 0.1)] kW")
+	if(!inserted_cylinder)
+		to_chat(user, "There is no piston cylinder inserted into \the [src].")
+
+/obj/machinery/atmospherics/binary/stirling/attackby(var/obj/item/W, var/mob/user)
+	if((istype(W, /obj/item/tank/stirling)))
+		if(inserted_cylinder)
+			return
+		if(!user.unEquip(W, src))
+			return
+		to_chat(user, SPAN_NOTICE("You insert \the [W] into \the [src]."))
+		inserted_cylinder = W
+		update_icon()
+		return TRUE
+	
+	if(IS_CROWBAR(W) && inserted_cylinder)
+		inserted_cylinder.dropInto(get_turf(src))
+		to_chat(user, SPAN_NOTICE("You remove \the [inserted_cylinder] from \the [src]."))
+		inserted_cylinder = null
+		stop_engine()
+		return TRUE
+
+	if(panel_open && IS_WRENCH(W))
+		var/target_frequency = input(user, "Enter the cycle frequency you would like \the [src] to operate at ([MAX_FREQUENCY/4] - [MAX_FREQUENCY] Hz)", "Stirling Frequency", cycle_frequency) as num | null
+		if(!CanPhysicallyInteract(user) || !target_frequency)
+			return
+		cycle_frequency = round(clamp(target_frequency, MAX_FREQUENCY/4, MAX_FREQUENCY))
+		to_chat(usr, SPAN_NOTICE("You adjust \the [src] to operate at a frequency of [cycle_frequency] Hz."))
+		return TRUE
```

Comment:
There are going to be part interaction conflicts with using crowbar and wrench here. Both are fine if panel is closed.

Reply:
Addressed, changed to panel closed.